src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java

```diff
@@ -269,59 +271,46 @@ public static String dropFirstAndLastChar(String s) {
     }
 
-    /**
-     * Parses a method descriptor string, and return a list of field descriptor
-     * strings, return type first, then parameter types
-     *
-     * @param descriptor the descriptor string
-     * @return the list of types
-     * @throws IllegalArgumentException if the descriptor string is not valid
-     */
-    public static List<ClassDesc> parseMethodDescriptor(String descriptor) {
-        int cur = 0, end = descriptor.length();
-        ArrayList<ClassDesc> ptypes = new ArrayList<>();
-        ptypes.add(null); // placeholder for return type
-
-        if (cur >= end || descriptor.charAt(cur) != '(')
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-
-        ++cur;  // skip '('
-        while (cur < end && descriptor.charAt(cur) != ')') {
-            int len = skipOverFieldSignature(descriptor, cur, end, false);
-            if (len == 0)
-                throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-            ptypes.add(resolveClassDesc(descriptor, cur, len));
-            cur += len;
-        }
-        if (cur >= end)
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-        ++cur;  // skip ')'
-
-        int rLen = skipOverFieldSignature(descriptor, cur, end, true);
-        if (rLen == 0 || cur + rLen != end)
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-        ptypes.set(0, resolveClassDesc(descriptor, cur, rLen));
-        return ptypes;
+    public static PrimitiveClassDescImpl forPrimitiveType(String descriptor, int offset) {
+        return switch (descriptor.charAt(offset)) {
+            case JVM_SIGNATURE_BYTE    -> CD_byte;
+            case JVM_SIGNATURE_CHAR    -> CD_char;
+            case JVM_SIGNATURE_FLOAT   -> CD_float;
+            case JVM_SIGNATURE_DOUBLE  -> CD_double;
+            case JVM_SIGNATURE_INT     -> CD_int;
+            case JVM_SIGNATURE_LONG    -> CD_long;
+            case JVM_SIGNATURE_SHORT   -> CD_short;
+            case JVM_SIGNATURE_VOID    -> CD_void;
+            case JVM_SIGNATURE_BOOLEAN -> CD_boolean;
+            default -> throw badMethodDescriptor(descriptor);
+        };
     }
 
-    private static ClassDesc resolveClassDesc(String descriptor, int start, int len) {
+    static ClassDesc resolveClassDesc(String descriptor, int start, int len) {
         if (len == 1) {
-            return Wrapper.forPrimitiveType(descriptor.charAt(start)).basicClassDescriptor();
+            return forPrimitiveType(descriptor, start);
         }
-        // Pre-verified in parseMethodDescriptor; avoid redundant verification
+
+        // objectDesc appears a lot during the bootstrap process, so optimize it
+        String objectDesc = "Ljava/lang/Object;";
+        if (len == objectDesc.length() && descriptor.regionMatches(start, objectDesc, 0, len)) {
+            return ReferenceClassDescImpl.CD_Object;
```

Comment:
Is there a reason we cannot use `ConstantDescs.CD_Object` and have to redefine it in `ReferenceClassDescImpl`? I see no dependency in initialization order, and the field in RCDI can be removed.